File: ClientApp/src/app/models/business/world-object.model.ts
import { signal, WritableSignal } from "@angular/core";
import { WorldObjectResponse } from "@models/response";
import { WorldObjectType } from "./world-object-type.model";

class WorldObject {
  type!: WorldObjectType;
  id: string;
  name: WritableSignal<string>;
  parentId: WritableSignal<string | null>;
  previousId: WritableSignal<string | null>;
  previewImageUrl: WritableSignal<string | null>;

  constructor(worldObjectDto: WorldObjectResponse) {
    this.id = worldObjectDto.id;
    this.name = signal(worldObjectDto.name);
    this.parentId = signal(worldObjectDto.parentId)
    this.previousId = signal(worldObjectDto.previousId);
    this.previewImageUrl = signal(worldObjectDto.previewImageUrl);
  }
}

export class WorldObjectFolder extends WorldObject {
  constructor(worldObjectDto: WorldObjectResponse) {
    super(worldObjectDto);
    this.type = WorldObjectType.Folder;
  }

  public Copy(id: string) {
    return new WorldObjectFolder({ type: this.type, name: this.name(), id, previewImageUrl: this.previewImageUrl(), parentId: this.parentId(), previousId: this.previousId() });
  }
}

export class WorldObjectCharacter extends WorldObject {
  isFavourite: WritableSignal<boolean>;

  constructor(worldObjectDto: WorldObjectResponse, isFavourite: boolean) {
    super(worldObjectDto);
    this.type = WorldObjectType.CharacterSheet;
    this.isFavourite = signal(isFavourite);
  }

  public Copy(id: string) {
    return new WorldObjectCharacter({ type: this.type, name: this.name(), id, previewImageUrl: this.previewImageUrl(), parentId: this.parentId(), previousId: this.previousId() }, this.isFavourite());
  }
}
export class WorldObjectHandout extends WorldObject {
  isFavourite: WritableSignal<boolean>;

  constructor(worldObjectDto: WorldObjectResponse, isFavourite: boolean) {
    super(worldObjectDto);
    this.type = WorldObjectType.Handout;
    this.isFavourite = signal(isFavourite);
  }

  public Copy(id: string) {
    return new WorldObjectCharacter({ type: this.type, name: this.name(), id, previewImageUrl: this.previewImageUrl(), parentId: this.parentId(), previousId: this.previousId() }, this.isFavourite());
  }
}

export type AnyWorldObject = WorldObjectFolder | WorldObjectCharacter | WorldObjectHandout;

File: Models/WorldObject/WorldObjectModel.cs
using qDshunUtilities.EF.Entities.WorldObjects;
using qDshunUtilities.Utils;

namespace qDshunUtilities.Models.WorldObject;


public class WorldObjectModel(WorldObjectEntity wo)
{
    public WorldObjectType Type { get; set; } = wo switch
    {
        FolderEntity => WorldObjectType.Folder,
        CharacterSheetEntity => WorldObjectType.CharacterSheet,
        HandoutEntity => WorldObjectType.Handout,
        _ => throw new NotImplementedException(),
    };
    public Guid Id { get; set; } = wo.Id;
    public string Name { get; set; } = wo.Name;
    public Guid? ParentId { get; set; } = wo.ParentId;
    public Guid? PreviousId { get; set; } = wo.PreviousId;
    public string PreviewImageUrl { get; set; } = wo?.PreviewImageUrl;

}
